File: MembershipService/MembershipService/Models/Membership.cs
using MembershipService.Models.Enums;

namespace MembershipService.Models;

public class Membership
{
    public Guid MembershipId { get; set; }

    public Guid UserId { get; set; }

    public Guid PackageId { get; set; }

    public DateTime StartDate { get; set; }

    public DateTime EndDate { get; set; }

    public DateTime CreatedDate { get; set; } = DateTime.UtcNow;

    public DateTime? CancelledDate { get; set; }

    public MembershipStatus Status { get; set; }

    public Package? Package { get; set; }

    public List<Checkin> Checkins { get; set; } = new();

    // Methods
    public bool IsActive()
    {
        return Status == MembershipStatus.Active && EndDate > DateTime.UtcNow;
    }

    public bool Renew(int durationDays)
    {
        if (Status == MembershipStatus.Cancelled)
            return false;

        EndDate = EndDate.AddDays(durationDays);
        Status = MembershipStatus.Active;
        return true;
    }

    public bool Cancel()
    {
        if (Status == MembershipStatus.Cancelled)
            return false;

        Status = MembershipStatus.Cancelled;
        CancelledDate = DateTime.UtcNow;
        return true;
    }

    public int DaysUntilExpiration()
    {
        var days = (EndDate - DateTime.UtcNow).Days;
        return days < 0 ? 0 : days;
    }

    public bool CanBeModified()
    {
        return Status != MembershipStatus.Cancelled && (StartDate > DateTime.UtcNow || EndDate < DateTime.UtcNow);
    }
}

File: frontend/src/types/membership.ts
export const MembershipStatus = {
    Active: 1,
    Expired: 2,
    Suspended: 3,
    Cancelled: 4
} as const;

export type MembershipStatus = (typeof MembershipStatus)[keyof typeof MembershipStatus];

export interface UserMembershipDto {
    membershipId: string;
    userId: string;
    packageId: string;
    startDate: string;
    endDate: string;
    createdDate: string;
    cancelledDate: string | null;
    status: MembershipStatus;
}

export interface MembershipPackageDto {
    packageId: string;
    name: string;
    description: string;
    price: number;
    duration: number;
    services: string[];
}
